frontend/catalyst/device/qjit_device.py

```diff
@@ -472,6 +472,9 @@ def preprocess(
         # measurement transforms may change operations on the tape to accommodate
         # measurement transformations, so must occur before decomposition
         measurement_transforms = self._measurement_transform_program()
+        if config is qml.devices.DefaultExecutionConfig:
+            # Do not modify DefaultExecutionConfig; copy into new config and modify the copy
+            config = deepcopy(config)
```

Comment:
Why not use `dataclasses.replace` instead of introducing new logic? I think this will also set up nicely for the change pennylane is about to make, since once the config is immutable making a copy of it will still result in an immutable object.

Reply:
I'm trying to avoid making modifications to attributes of `DefaultExecutionConfig` that are mutable objects, e.g. dictionaries, since `replace` returns a shallow copy of the dataclass. Here's an example:

```python
from copy import deepcopy
from dataclasses import dataclass, field, replace, FrozenInstanceError

@dataclass(frozen=True)
class Foo:
    params: dict = field(default_factory=dict)


default_foo = Foo()

try:
    default_foo.params = {}
except FrozenInstanceError as err:
    print(f"Caught FrozenInstanceError: {err}")

foo1 = replace(default_foo, params={'x': 1})
print(f"default_foo: {default_foo}")
print(f"foo1:        {foo1}\n")

foo2 = replace(default_foo)
foo2.params['x'] = 2
print(f"default_foo: {default_foo}")
print(f"foo2:        {foo2}\n")

print("Resetting default_foo...")
default_foo = Foo()

foo3 = deepcopy(default_foo)
foo3.params['x'] = 3
print(f"default_foo: {default_foo}")
print(f"foo3:        {foo3}")
```

Output:

```
Caught FrozenInstanceError: cannot assign to field 'params'
default_foo: Foo(params={})
foo1:        Foo(params={'x': 1})

default_foo: Foo(params={'x': 2})
foo2:        Foo(params={'x': 2})

Resetting default_foo...
default_foo: Foo(params={})
foo3:        Foo(params={'x': 3})
```

Quick summary:

1. We can't assign to `params` for `default_foo`, as expected since it's frozen.
2. `foo1`: using `replace` and supplying a new dictionary does not modify `default_foo`, but overwrites any existing values in its `params`.
3. `foo2`: using `replace` without additional arguments creates a shallow copy; modifying `params` in `foo2` also modifies `params` in `default_foo` (not good!)
4. `foo3`: using `deepcopy` it's possible to modify `params` without touching `default_foo`.

It appears that the immutability of a dataclass is lost after both a `replace` and a `deepcopy`, at least when using `@dataclass(frozen=True)`. We could use `replace` and supply a new dictionary, but this maybe isn't ideal if `DefaultExecutionConfig` ever contains any items in its dictionaries since these would be overwritten.

Reply:
Wow the more you know :o Nice investigating! 

I think I'm okay with this then, it's hard to guarantee right now that it will work with the way upstream will make the class immutable, but I guess we'll notice immediately if something breaks 👍

Reply:
Although I think I would feel better if we just made a copy unconditionally, the id check against DefaultExecutionConfig does feel a little brittle 🤔

Reply:
That would probably be for the best, checking if the config is the `DefaultExecutionConfig` may be a case of premature optimization; it would be much safer to just make a copy. The overhead should be reasonably small given that `ExecutionConfig` is a fairly small class.

Would your preference be to use the

```python
def preprocess(config=None):
    if config is None:
        config = ExecutionConfig()
```

pattern, or make deep copies of `DefaultExecutionConfig`? The former will be safer, since right now we can't guarantee that `DefaultExecutionConfig` is immutable, so it's possible that we end up deep-copying it in a modified state.

Reply:
That is a good actually since we decided yesterday that PL will update the signature to this in the near future 👍